### ISSUE TITLE
BRGenomics

### ISSUE THREAD

**Issue by mdeber** [Bioconductor/Contributions#1398]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/mdeber/BRGenomics

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by LiNk-NY** [comment]
Hi Mike, @mdeber 
Are you still actively developing the package?
If so, the submission may be a bit too early.
Please advise.
-Marcel

**Comment by mdeber** [comment]
Hi Marcel, @LiNk-NY 
Sorry about all the late feature additions – some of those pushes were battling the windows build system, but I did unexpectedly end up adding features and further optimizing a few functions, based on user feedback. I'm going to push _one more_ small feature addition to one of the functions right now, but aside from that small change things should be settled. I and a few others have been using it for a while now and I'm not planning on changing anything else.
-Mike

**Comment by mdeber** [comment]
Hi Marcel, @LiNk-NY
Is my package suitable for review? Again, sorry for the late feature additions, but let me know.
Mike

**Comment by mdeber** [comment]
Hi Marcel, @LiNk-NY 
I hope you're well. Is there any action I should take to have my package reviewed? Since there are several papers in preparation that have used BRGenomics, I'd really like to get an application note published soon. If there are any changes that should be made, I'd like to make them before the application note is finalized, and while other users have time to modify their code (if necessary) before publication. I understand that I'm graciously dependent on the pure generosity of busy academics, and I'd appreciate any kind of update or advice concerning the process, e.g. if I should submit a new issue or ask the mailing list if anyone is willing to participate.
Thank you in advance!

**Comment by LiNk-NY** [comment]
Hi Mike, @mdeber 

Thank you for your patience. Package submissions close to the release
deadline should expect longer than usual review times. 
Your package is next in the queue and should not take more than a
couple of days.

Best,
Marcel

**Comment by mdeber** [comment]
Ok thanks – I greatly appreciate it!
Mike

**Comment by LiNk-NY** [comment]
Hi Mike, @mdeber 
Thank you for your submission. 
Please see below for the review. 
Overall, the package looks in good condition. 

Best,
Marcel

--- 

# BRGenomics #1398

## DESCRIPTION

* It looks good.
* Add a `BugReports` field.

## NAMESPACE

* It looks good.
* Try to make function names such as `aggregateByNdimensionalBins`,
`densityInNdimensionalBins`, etc. shorter (if possible)
* If you are using `SummarizedExperiment` as a standard representation in
the package it may be better to import the entire package rather than just
the `SummarizedExperiment` function.

## vignettes

* The front matter looks great.
* (optional) The text width should not exceed 80 columns as it's easier to read,
edit, and maintain.
* Run vignette through spell checking software
* Update the installation instructions to make use of `BiocManager::install`
* Remove mention of `Rtools` since binaries would be provided for Windows users
* It is quite lengthy and thorough, consider splitting the vignette into
multiple parts to thematically target users

## R

* I think that functions should not be greedy in terms of core-usage, they
should use options set by the user (e.g., as `options("ncpus")`) instead of
`detectCores()`
* Are there existing packages that import these type of files already (e.g.,
`bam`, `bigwig`)? If not, perhaps differentiate these functions by giving them
a more descriptive name.
* Use identical to test equality between strings instead of `==`, it's not as
robust with edge cases and multiple input vectors (`fix.start` and `fix.end`)
in `genebodies`.
* Code that intersects and reduces ranges by gene matches the functionality
in `RaggedExperiment` package. Perhaps you may want to look into that data
structure?

**Comment by LiNk-NY** [comment]
Hi Mike, @mdeber 
Please keep in mind that the response to the review and the changes are due days before the April 22nd deadline: https://www.bioconductor.org/developers/release-schedule/. This will give me some time to look over the changes. Thank you. 
-Marcel

**Comment by mdeber** [comment]
Thanks for the heads up. I'll try to address all of your comments as soon as possible.
Thanks again, 
-Mike

**Comment by mdeber** [comment]
Hey Marcel, @LiNk-NY 
I think I've addressed just about everything. I went through point-by-point, sometimes too wordy, so sorry about that. 
Thanks a lot for everything, and let me know if anything has to change.

-Mike

> # BRGenomics #1398
> ## DESCRIPTION
> 
>     * It looks good.
> 
>     * Add a `BugReports` field.
_Done_
> 
> 
> ## NAMESPACE
> 
>     * It looks good.
> 
>     * Try to make function names such as `aggregateByNdimensionalBins`,
>       `densityInNdimensionalBins`, etc. shorter (if possible)
_Done. Definitely for the better_
> 
>     * If you are using `SummarizedExperiment` as a standard representation in
>       the package it may be better to import the entire package rather than just
>       the `SummarizedExperiment` function.
Thanks for the tip, but I currently only use one as a quick ~required stepping stone for the DESeq2 functionality, so I think I'll leave this as-is
> 
> ## vignettes
> 
>     * The front matter looks great.
> 
>     * (optional) The text width should not exceed 80 columns as it's easier to read,
>       edit, and maintain.
_Done - hadn't been doing in the vignettes/readme, but it is nicer now that I have_
> 
>     * Run vignette through spell checking software
_Thanks, those typos did not look good_
>     * Update the installation instructions to make use of `BiocManager::install`
_Done_
>     * Remove mention of `Rtools` since binaries would be provided for Windows users
I've clarified the installation instructions to future-proof for the Bioc release, but I left the instructions for installing from Github underneath
>     * It is quite lengthy and thorough, consider splitting the vignette into
>       multiple parts to thematically target users
_Yes, it was out of hand. Much improved now_
> 
> ## R
> 
>     * I think that functions should not be greedy in terms of core-usage, they
>       should use options set by the user (e.g., as `options("ncpus")`) instead of
>       `detectCores()`
_Done. Perfectly sensible_
>     * Are there existing packages that import these type of files already (e.g.,
>       `bam`, `bigwig`)? If not, perhaps differentiate these functions by giving them
>       a more descriptive name.
_Sorry for the text, tl;dr I agree and I'm trying to figure out how best to name/organize this functionality_

This one I'm still struggling a bit with, because I agree with you completely: rtracklayer has functions `import.bw` and `import.bedGraph`, and I've written here `import_bigWig` and `import_bedGraph`. There's too much dissonance there, but I'll need to think more about how to change things because these functions are very convenient, and I am planning on making them a even a bit more automated in terms of automatically importing stranded data.

(On the topic of their value, bigWigs and bedGraphs are common and efficient, but not stranded, and I've noticed it's surprisingly challenging for novices to import a lot of them efficiently because the files have to be paired. More often than not I see this is done with huge text bombs of all the file paths and file names, and unfortunately loading complete bigWigs into memory remains surprisingly CPU-bound and it benefits immensely from parallelization. But more importantly, and a motivation of this package, is that it's very common for people working with basepair-resolution ("spikey") to fail to account for the run-length encoding in quantifications).

Anyway I'll think more about this, but I agree they should be renamed, or the workflow should be adjusted somehow...

On the other hand, `import_bam` is more involved, and something I had to spend some time on because I really couldn't find a direct drop-in replacement. Under my testing conditions a few other implementations broke (all through `bpiterate`), which may be a temporary bug. However, the integration of post-importation processing steps (particularly when it comes to paired-end sequencing with basepair-resolution processing) creates the opportunity for huge performance gains in the future, once I get around to implementing it

>     * Use identical to test equality between strings instead of `==`, it's not as
>       robust with edge cases and multiple input vectors (`fix.start` and `fix.end`)
>       in `genebodies`.
I added in `match.arg` for these, after which point I kept the `==` for the sake of readability. Let me know if this isn't a solution
>     * Code that intersects and reduces ranges by gene matches the functionality
>       in `RaggedExperiment` package. Perhaps you may want to look into that data
>       structure?
Will do. I haven't gotten to it yet but the package sounds interesting. For now I'll keep these fairly basic functions in place, but I'll see if there isn't a more complete/useful direction to point users for these kinds of manipulations.

Thanks again Marcel

**Comment by LiNk-NY** [comment]
Hi Mike, @mdeber 

Thanks for your thorough response. 
I've had a second look at the changes and it's looking good!

I think a name change would be good especially if you are providing
an extra layer of convenience to your users.
Check on the `RaggedExperiment` representation as you further develop
your package. 
<sup> Disclaimer: I am one of the developers of `RaggedExperiment`. </sup>

Thank you for your submission. Your package has been accepted. 
Best,
Marcel

**Comment by mdeber** [comment]
@LiNk-NY Wonderful! Thanks a lot.

As a side note, I'm surprised I hadn't come across RaggedExperiment before, because I was specifically looking for data structures for sparse data, and I hadn't had much luck using sparse matrices from the Matrix package (huge performance penalty, at least where tens- to low-hundreds of millions of read-positions are concerned).

Anyway, once again thanks, I'll play around with this for sure. All the best,
-Mike

**Comment by LiNk-NY** [comment]
The performance of `RaggedExperiment` with large and sparse matrices is untested. It may not be well suited for these representations (although it uses `GRangesList` internally) based on an open issue on the repository. It would be good to get your feedback if you encounter any issues. You can post them at https://github.com/Bioconductor/RaggedExperiment/issues.

**Comment by mdeber** [comment]
Great. Yea I checked the source code and it looks like the internal representation would be essentially the same as what I was informally calling "multiplexed GRanges" objects. (In hindsight would have been better to use this existing infrastructure). The issue I found with complex basepair-resolution data is you end up with huge ranges and mostly zeroes in the metadata, which leads to memory-bottlenecks that isn't offset by any improvement of having a single GRanges object. My 2 cents is that I think there's a growing amount of sparse genomics data out there, and if there are performance gains to be had, I think a fast, memory-efficient handling sparse multi-dimensional data would be very generally useful as a core Bioconductor functionality. (Although I did find personally the dgCMatrix to be catastrophic for performance and stability)

Anyway, just food for thought. Thanks again Marcel.

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/mdeber.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("BRGenomics")`. The package 'landing page' will be created at

https://bioconductor.org/packages/BRGenomics

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
